### PR TITLE
Add label selector to exclude deployments and cronjobs

### PR DESCRIPTION
**Pull request by dirk39**
Hello, with this PR we want to fix issue #172. We noticed that the behaviour of the Cronjob selector differs from the deployment. Do you think we can replicate the deployment behaviour in the cronjob? So we can add the matchLabels functionality.

[Open Source Saturday](https://www.meetup.com/it-IT/Open-Source-Saturday-Milano/)

### CODE REVIEW

**Comment by ludusrusso** [comment]
LGTM

**Review by davidebianchi — COMMENTED** [review]
Hi @dirk39, thank you for the PR. Great work!

I didn't know the Open Source Saturday meetup, very interesting!

Here a couple of comments.

The implementation of the CronJob differs because it uses the FieldSelector query, do you mean this? This should be more performant, as it permits to get only the resources needed. It is not possible to add this filter using the `FieldSelector`?
Otherwise, yes I suppose that it is possible to change the implementation to add this filter.

**Comment by dirk39** [comment]
Hi @davidebianchi , the general idea was to uniform the behavior of the deployment and cronjob filter. I think that the deployment solution is easier to maintain but if you prefer I can have a look to the fieldSelector to see if they support the labelMatch.

File: config/crd/bases/kube-green.com_sleepinfos.yaml [hunk 1]
```diff
@@ -47,13 +47,18 @@ spec:
...
+                      description: MatchLabels which identify the kubernetes resource
+                        by labels
+                      type: object
                     name:
                       description: Name which identify the kubernetes resource.
                       type: string
                   required:
                   - apiVersion
                   - kind
-                  - name
```
[the remark this answers is not in the document]
Comment (dirk39) [reply config/crd/bases/kube-green.com_sleepinfos.yaml:56]: I have to see. One solution might be to validate the sleepinfo checking that name or matchLabels must be set.

**Comment by davidebianchi** [comment]
CronJobs have been made after Deployments, so when I developed the Deployments I didn't know the fieldSelector. But with the labels, there might be a large number of Deployments and CronJobs to exclude so it could be better to use it. So yes, if you want to see for the support it would be better, but also this implementation is ok for me.

File: controllers/sleepinfo/deployments/deployments.go [hunk 2]
```diff
@@ -108,24 +108,46 @@ func (d deployments) getListByNamespace(ctx context.Context, namespace string) (
...
+		if labelMatch(deployment.Labels, exclusion.MatchLabels) {
+			return true
 		}
 	}
-	return excludedDeploymentName
+
+	return false
+}
+
+func labelMatch(labels, matchLabels map[string]string) bool {
```
[the remark this answers is not in the document]
Comment (dirk39) [reply controllers/sleepinfo/deployments/deployments.go:136]: Done

File: controllers/sleepinfo/deployments/deployments_test.go [hunk 3]
```diff
@@ -476,6 +510,55 @@ func TestDeploymentOriginalReplicas(t *testing.T) {
...
+			labels: map[string]string{
+				"foo-key": "foo-value",
+			},
+			matchLabels: map[string]string{
+				"app-key": "app-value",
+			},
+			expected: false,
+		},
+		{
+			name: "Match sucess",
```
Comment (davidebianchi) [inline controllers/sleepinfo/deployments/deployments_test.go:542]:
```suggestion
			name: "Match success",
```

**Comment by davidebianchi** [comment]
Hi @dirk39, is it the introduction of labelSelector finished? I see that the last commit contains is in wip.

**Comment by dirk39** [comment]
Hi @davidebianchi , I've introduced the labelselector for the cronjob but I have to add tests to validate the solution.

**Comment by dirk39** [comment]
Hi @davidebianchi, I'm sorry for my delay. I've added unit and e2e tests for the CJ. Let me know if it looks good to you

**Review by davidebianchi — APPROVED** [review]
LGTM!

Thanks for the contribution, great work!

**Pull request merged**